### PR TITLE
Fix O(n^2) complexity in the watcher hidden within nested loops created by table.{insert,remove} 

### PR DESCRIPTION
**Pull request by dmitmel**
Ref: #12 \
Depends on: #17

Addresses point 6 of my original PR. Also, the reproduction procedure I have provided is wrong, here is a better one:

1. Grab a file with many lines (<https://github.com/microsoft/TypeScript/blob/v4.4.3/lib/tsserver.js> I mentioned in that PR will do)
2. Jump to around 80% of the file (you can literally type `80%` in the normal mode)
3. Do `d10gg` (you can also try `u` afterwards)
4. Catastrophic lag!

### CODE REVIEW

File: lua/cmp_buffer/buffer.lua [hunk 1]
```diff
@@ -68,82 +72,102 @@ end
...
 
-  self.words[i] = words
+  self.lines_words[linenr] = words
 end
 
 --- get_words
 function buffer.get_words(self)
   local words = {}
-  for _, line in ipairs(self.words) do
+  for _, line in ipairs(self.lines_words) do
```
Comment (hrsh7th) [inline lua/cmp_buffer/buffer.lua:165]: lines_words[i] can be `vim.NIL`, I think.
So we should check `type(line) == 'table'` here.

**Comment by hrsh7th** [comment]
diff https://github.com/hrsh7th/cmp-buffer/compare/main...dmitmel:optimize-hidden-nested-loops

File: lua/cmp_buffer/buffer.lua (hunk 1 above)
Comment (dmitmel) [reply lua/cmp_buffer/buffer.lua:165]: No, vim.NIL values are used for extending the internal array part of the `lines_words` table. After loops in the branch with `append` are done, there are no vim.NIL values in `lines_words`.

**Comment by dmitmel** [comment]
Rebased

File: lua/cmp_buffer/buffer.lua (hunk 1 above)
Comment (dmitmel) [reply lua/cmp_buffer/buffer.lua:165]: Well, I guess the loops can be made clearer by rewriting the variable `delta` as `new_last_line - old_last_line`, but I don't have a better argument than "trust me"

File: lua/cmp_buffer/buffer.lua [hunk 2]
```diff
@@ -68,32 +72,64 @@ end
...
+          self.lines_words[i] = {}
+        end
+      elseif delta < 0 then -- remove
+        -- Move backwards the unchanged elements in the tail part.
+        for i = old_last_line + 1, self.lines_count do
+          self.lines_words[i + delta] = self.lines_words[i]
+        end
+        -- Remove (already copied) tables from the end, in reverse order, so
+        -- that we don't make holes in the lines table.
+        for i = self.lines_count, new_lines_count + 1, -1 do
```
Comment (hrsh7th) [inline lua/cmp_buffer/buffer.lua:117]: This +1 seems unnecessary, I think.

Comment (hrsh7th) [reply lua/cmp_buffer/buffer.lua:117]: Ah. I understand. It's necessary.

**Comment by hrsh7th** [comment]
LGTM.

File: lua/cmp_buffer/buffer.lua (hunk 2 above)
Comment (dmitmel) [reply lua/cmp_buffer/buffer.lua:117]: Yeah, the point of this loop is to remove extra lines at the end.

**Pull request merged**